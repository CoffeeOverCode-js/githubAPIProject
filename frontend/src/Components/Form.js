import React, { useState } from 'react';

const Form = ({ username, setUsername, userJSON, setUserJSON, error, setError, repoJSON, setRepoJSON, repoNames, setRepoNames, commitJSON, setCommitJSON, limitedRepos, setLimitedRepos  }) => {


    async function fetchData() {
        setCommitJSON({})
        try {
            // Fetch user data
            const userResponse = await fetch(`http://localhost:8080/user/${username}`);
            if (!userResponse.ok) {
                setError(true);
                setUserJSON({ message: "User Not Found" });
                setCommitJSON({message: "User Not Found"})
                throw new Error(`HTTP error! Status: ${userResponse.status}`);
            }
            setError(false);
            const userData = await userResponse.json();
            setUserJSON(userData);
    
            // Fetch repository data
            const repoResponse = await fetch(`http://localhost:8080/repo/${username}`);

            if (!repoResponse.ok) {
              setError(true);
              setRepoJSON({ message: "Repositories Not Found" });
              throw new Error(`HTTP error! Status: ${repoResponse.status}`);
            }
            
            setError(false);
            const repoData = await repoResponse.json();
            if (repoData.length >= 20) {
                repoData.slice(0, 20);
            }
            // Slice the repoData to include only the first 5 repositories
            const slicedRepoData = repoData.slice(0, 4);
            
            setRepoJSON(repoData);
            setLimitedRepos(slicedRepoData)
            
    
            // Extract repository names
            const repoNames = repoData.map(repo => repo.name);
            setRepoNames(repoNames)
            const commitObj = [];
            
            // Loops through the slicedRepoData to find all the commits of the different repos
            for (let i = 0; i < slicedRepoData.length; i++) {
                const commitResponse = await fetch(`https://api.github.com/repos/${username}/${repoNames[i]}/commits`
                );
                
                if (!commitResponse.ok) {
                    setError(true);
                    setCommitJSON({ message: "Commits Not Found" });
                    throw new Error(`HTTP error! Status: ${commitResponse.status}`);
                }
    
                setError(false);
                const commitData = await commitResponse.json();
                commitObj.push(commitData[0]);
            }
            // Sets the relevant data
            setCommitJSON(commitObj);
        } catch (e) {
            console.error('Error:', e);
        }
    }
    


    // Function that calls the fetch data when Submit button is clicked
    const handleSubmit = (e) => {
        e.preventDefault();
        fetchData();
    };

    // Returns the HTML Layout of the component
    return (
        <div>
            <h1 className='display-4'>Form</h1>
            <form onSubmit={handleSubmit}>
                <label>Enter username:</label>
                <input
                    type="text"
                    name="username"
                    value={username}
                    onChange={(e) => setUsername(e.target.value)}
                    required
                />
                <button type="submit" className='btn btn-default'>Submit</button>
            </form>
        </div>
    );
};

export default Form;
