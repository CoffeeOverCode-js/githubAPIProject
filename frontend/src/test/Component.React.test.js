// Imports the React library for JSX rendering
import React from "react";
import renderer from 'react-test-renderer';
// Imports component to be tested
import Form from "../Components/Form";

// Tests to see if the content renders correctly
test('renders correctly', () => {
    // Creates a snapshot of the form and converts it to JSON
    const tree = renderer
        .create(<Form />)
        .toJSON();
        // Compares the rendered component with the snapshot
    expect(tree).toMatchSnapshot();
});