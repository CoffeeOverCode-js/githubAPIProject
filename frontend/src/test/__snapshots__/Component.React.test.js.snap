// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1
    className="display-4"
  >
    Form
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label>
      Enter username:
    </label>
    <input
      name="username"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <button
      className="btn btn-default"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
